File: 46-permutations/permutations.java
class Solution {
    private void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
    private void permuteUtil(int[] nums, int idx, List<List<Integer>> ans){
        if(idx==nums.length-1){
            List<Integer> res= new ArrayList<>();
            for(int num: nums) res.add(num);
            ans.add(res);
            return;
        }
        for(int j=idx;j<nums.length;j++){
            swap(nums, idx, j);
            permuteUtil(nums, idx+1, ans);
            swap(nums, idx, j);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        permuteUtil(nums, 0, ans);
        return ans;
    }
}

File: 46-permutations/permutations.cpp
class Solution {
public:
    void permuteUtil(vector<int> arr, int idx, vector<vector<int>> &ans){
        if(idx==arr.size()){
            ans.push_back(arr);
            return;
        }
        for(int j=idx;j<arr.size();j++){
            vector<int> curr(arr);
            swap(curr[j],curr[idx]);
            permuteUtil(curr, idx+1, ans);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        permuteUtil(nums, 0, ans);
        return ans;
    }
};